src/com/phonebookservice/util/MyArrayList.java

```diff
@@ -0,0 +1,416 @@
+                this.data[i] = this.data[i - 1];
+            }
+
+            this.data[index] = element;
+            this.index++;
+        } else {
+            this.add(element);
+        }
+    }
+
+    /**
+     * Appends all of the elements in the collection to the end of this list, in
+     * the order that they are returned by the collection's iterator .
+     *
+     * @param collection the collection.
+     * @return true if added else false.
+     */
+    @Override
+    public boolean addAll(final Collection<? extends T> collection) {
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * Inserts all of the elements in the collection into this list at the
+     * specified position.
+     *
+     * @param index      the index.
+     * @param collection the collection.
+     * @return true if added else false.
+     */
+    @Override
+    public boolean addAll(final int index,
+            final Collection<? extends T> collection) {
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * Removes all of the elements from this list.
+     */
+    @Override
+    public void clear() {
+        for (int i = 0; i < this.index; i++) {
+            this.data[i] = null;
+        }
+
+        this.index = 0;
+    }
+
+    /**
+     * Checks if this list contains the specified element.
+     *
+     * @param element the element.
+     * @return true if this list contains the specified element.
+     */
+    @Override
+    public boolean contains(final Object element) {
+        return this.lastIndexOf(element) != -1;
+    }
+
+    /**
+     * Get the element at the specified position in this list..
+     *
+     * @param index the index.
+     * @return the element at the specified position in this list.
+     */
+    @Override
+    public T get(final int index) {
+        this.checkIndex(index);
+        return this.data[index];
+    }
+
+    /**
+     * Returns the index of the first occurrence of the specified element in
+     * this list, or -1 if this list does not contain the element.
+     *
+     * @param element the element
+     * @return the index
+     */
+    @Override
+    public int indexOf(final Object element) {
+        if (element == null) {
+            for (int i = 0; i < this.index; i++) {
+                if (this.data[i] == null) {
+                    return i;
+                }
+            }
+        } else {
+            for (int i = 0; i < this.index; i++) {
+                if (element.equals(this.data[i])) {
+                    return i;
+                }
+            }
+        }
+        return -1;
+    }
+
+    /**
+     * @param element the element.
+     * @return the index of the last occurrence of the specified element in this
+     *         list, or -1 if this list does not contain the element.
+     */
+    @Override
+    public int lastIndexOf(final Object element) {
+        if (element == null) {
+            for (int i = this.index - 1; i >= 0; i--) {
```

Comment:
Imagine the body of the for loop is big

For () {
Method1 
Method 2
Method 3
Method 4
}

You repeated for loop two time in the code. If you need to modify it, you need to modify in 2 for loop, one if element == null and other when element != null.


What I suggest here to have a private method 'equal(element, index)'

Private Boolean equal(T element, int index) {

Return element == this.arr[i] ||
    This.arr[i] != null && arr[i].equals(element)

}

Reply:
(Y)